src/pynwb/validation.py

```diff
@@ -0,0 +1,209 @@
+      - BuildManager object for opening the file for validation
+      - Dict with the full result from NWBHDF5IO.load_namespaces
+    """
+    catalog = NamespaceCatalog(NWBGroupSpec, NWBDatasetSpec, NWBNamespace)
+    namespace_dependencies = NWBHDF5IO.load_namespaces(catalog, path)
+
+    # determine which namespaces are the most specific (i.e. extensions) and validate against those
+    candidate_namespaces = set(namespace_dependencies.keys())
+    for namespace_dependency in namespace_dependencies:
+        candidate_namespaces -= namespace_dependencies[namespace_dependency].keys()
+
+    # TODO remove this workaround for issue https://github.com/NeurodataWithoutBorders/pynwb/issues/1357
+    candidate_namespaces.discard('hdmf-experimental')  # remove validation of hdmf-experimental for now
+    namespaces = sorted(candidate_namespaces)
+
+    if len(namespaces) > 0:
+        tm = TypeMap(catalog)
+        manager = BuildManager(tm)
+    else:
+        manager = None
+
+    return namespaces, manager, namespace_dependencies
+
+
+@docval({'name': 'path', 'type': HDMFIO, 'doc': 'the HDMFIO object to read from'},
+        {
+            'name': 'use_cached_namespaces',
+            'type': str, 'doc': 'Whether to use namespaces cached within the file for validation.',
+            'default': True
+        },
+        {
+            'name': 'namespace',
+            'type': Optional[str], 'doc': 'Whether to use namespaces cached within the file for validation.',
+            'default': None
+        },
+        returns="errors in the file", rtype=list,
+        is_method=False)
+def validate_file(**kwargs):
+    """Validate an NWB file against a namespace or its cached namespaces."""
+    path, use_cached_namespaces, namespace = getargs("path", "use_cached_namespaces", "namespace", kwargs)
+
+    if use_cached_namespaces:
+        catalog = NamespaceCatalog(NWBGroupSpec, NWBDatasetSpec, NWBNamespace)
+        ns_deps = NWBHDF5IO.load_namespaces(catalog, path)
+        s = set(ns_deps.keys())       # determine which namespaces are the most
+        for k in ns_deps:             # specific (i.e. extensions) and validate
+            s -= ns_deps[k].keys()    # against those
+        # TODO remove this workaround for issue https://github.com/NeurodataWithoutBorders/pynwb/issues/1357
+        if 'hdmf-experimental' in s:
+            s.remove('hdmf-experimental')  # remove validation of hdmf-experimental for now
+        namespaces = list(sorted(s))
+        if len(namespaces) > 0:
+            tm = TypeMap(catalog)
+            manager = BuildManager(tm)
+            specloc = "cached namespace information"
+        else:
+            manager = None
+            namespaces = [CORE_NAMESPACE]
+            specloc = "pynwb namespace information"
+            print("The file {} has no cached namespace information. "
+                  "Falling back to {}.".format(path, specloc), file=sys.stderr)
```

Comment:
Is this code necessary? Most of it was moved into `get_cached_namespaces_to_validate`